synapse/logging/context.py

```diff
@@ -751,6 +805,10 @@ def run_in_background(f, *args, **kwargs) -> defer.Deferred:
     # At this point we should have a Deferred, if not then f was a synchronous
     # function, wrap it in a Deferred for consistency.
     if not isinstance(res, defer.Deferred):
+        # `res` is not a `Deferred` and not a `Coroutine`.
+        # There are no other types of `Awaitable`s we expect to encounter in Synapse.
+        assert not isinstance(res, Awaitable)
+
```

Comment:
This turns out to be false. We have `DoneAwaitable`s from `maybe_awaitable` and completed `Future`s from `make_awaitable`.

I've added support for these in #12505.